File: src/backends/keychain/KeychainBackend.ts
import { Backend, GenericPasswordLocator, Password, PasswordLocator } from '~/types'
import { exec } from '~/util'
import { InvalidLocatorError, LocatorNotFoundError, UnsupportedPlatformError } from './errors'
import { parseKeychainOutput } from './parsing'

export default class KeychainBackend implements Backend {

  constructor() {
    checkPlatform()
  }

  public async listPasswords(): Promise<PasswordLocator[]> {
    const {stdout} = await exec(SECURITY_PATH, ['dump-keychain'], {
      capture: ['stderr', 'stdout'],
      throws:  true,
    })

    const lines = (stdout ?? '').split('\n')
    const items = parseKeychainOutput(lines)

    const locators: PasswordLocator[] = []
    for (const item of items) {
      const {svce, acct} = item.attributes
      if (svce !== SERVICE) { continue }
      if (acct == null) { continue }

      locators.push({type: 'generic', name: acct})
    }

    return locators
  }

  public async getPassword(locator: PasswordLocator): Promise<Password | null> {
    if (locator.type !== 'generic') {
      throw new InvalidLocatorError()
    }

    const {stderr} = await exec(SECURITY_PATH, [
      'find-generic-password',
      '-s', SERVICE,
      '-a', locator.name,
      '-g',
    ], {
      capture: ['stderr'],
      throws:  false,
    })

    const lines = (stderr ?? '').split('\n')
    for (const line of lines) {
      const match = line.match(/^password:\s(.+)$/)
      if (match == null) { continue }

      const passwordRaw = match[1].trim()
      if (passwordRaw.length === 0) { continue }

      return {
        locators: [locator],
        password: JSON.parse(passwordRaw),
      }
    }


    return null
  }

  public async storePassword(password: Password): Promise<boolean> {
    const genericLocator = PasswordLocator.get<GenericPasswordLocator>(password, 'generic')
    if (genericLocator == null) {
      throw new LocatorNotFoundError()
    }

    const {exitCode} = await exec(SECURITY_PATH, [
      'add-generic-password',
      '-s', SERVICE,
      '-a', genericLocator.name,
      '-w', password.password,
      '-U',
    ], {
      throws: true,
    })

    return exitCode === 0
  }

  public async removePassword(locator: PasswordLocator): Promise<boolean> {
    if (locator.type !== 'generic') {
      throw new InvalidLocatorError()
    }

    const {exitCode} = await exec(SECURITY_PATH, [
      'delete-generic-password',
      '-s', SERVICE,
      '-a', locator.name,
    ], {
      throws: false,
    })

    return exitCode === 0
  }

}

function checkPlatform() {
  if (process.platform !== 'darwin') {
    throw new UnsupportedPlatformError()
  }
}

const SECURITY_PATH = '/usr/bin/security'
const SERVICE = 'ypass'

File: python/ypass/backends/keychain/keychain_backend.py
import subprocess

from ...util import PasswordExists, PasswordNotFound, PasswordQuery
from .consts import SERVICE_NAME
from .password_reader import PasswordReader


class KeychainBackend:

  def list(self, query: PasswordQuery, include_passwords = False):
    cmd = ['/usr/bin/security', 'dump-keychain']
    
    process = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    if process.stdout is None: return []

    reader   = PasswordReader(process.stdout, process.stderr)
    filtered = (password for password in reader() if query.match(password))

    if include_passwords:
      return (
        self._retrieve(PasswordQuery(name = password.name))
        for password in filtered
      )
    else:
      return filtered


  def show(self, query: PasswordQuery):
    password = self._retrieve(query)
    if not password: raise PasswordNotFound
    return password

  def has(self, query: PasswordQuery):
    pass

  def add(self, name: str, password: str):
    exitcode = self._store(name, password, False)
    if exitcode == 45: raise PasswordExists

  def update(self, name: str, password: str):
    exitcode = self._store(name, password, True)
    print(exitcode)
    if exitcode == 44: raise PasswordNotFound

  def delete(self, name: str):
    cmd = [
      '/usr/bin/security', 'delete-generic-password',
      '-s', SERVICE_NAME,
      '-a', name,
    ]

    process = subprocess.run(cmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
    match process.returncode:
      case 44: raise PasswordNotFound
      
  def _retrieve(self, query: PasswordQuery):
    cmd = [
      '/usr/bin/security', 'find-generic-password',
      '-s', SERVICE_NAME,
    ]

    if query.name:
      cmd += ['-a', query.name]

    cmd += ['-g']

    process = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    if process.stdout is None: return []

    reader    = PasswordReader(process.stdout, process.stderr)
    generator = (password for password in reader())
    return next(generator, None)

  def _store(self, name: str, password: str, update: bool):
    cmd = [
      '/usr/bin/security', 'add-generic-password',
      '-s', SERVICE_NAME,
      '-a', name,
    ]

    if update:
      cmd += ['-U']

    cmd += ['-w', password]

    process = subprocess.run(cmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
    return process.returncode